File: src/resources/technical analysis/trend/simple moving average/simpleMovingAverage.kt
import kotlin.math.pow

data class SMAResult(
    val period: Int,
    val averages: Sequence<Double>
)

fun calculateSMA(prices: List<Double>, period: Int): Sequence<Double> = sequence {
    require(period > 0) { "Period must be a positive integer." }
    require(prices.size >= period) { "Prices list length must be greater than or equal to the period." }

    val buffer = DoubleArray(period)
    var sum = 0.0
    var index = 0

    // Initialize the sum with the first 'period' prices
    for (i in 0 until period) {
        buffer[i] = prices[i]
        sum += prices[i]
    }

    // Yield the initial SMA
    yield(sum / period)

    // Calculate the rest of the SMAs using the efficient update formula
    for (i in period until prices.size) {
        sum = sum - buffer[index] + prices[i]
        buffer[index] = prices[i]
        index = (index + 1) % period

        if (period.isPowerOfTwo()) {
            // If period is a power of 2, use bit shift for division
            yield(sum * (1.0 / period))
        } else {
            yield(sum / period)
        }
    }
}

fun getSMAResult(prices: List<Double>, period: Int): SMAResult {
    return SMAResult(
        period = period,
        averages = calculateSMA(prices, period)
    )
}

// Extension function to check if a number is a power of 2
fun Int.isPowerOfTwo(): Boolean = this > 0 && (this and (this - 1)) == 0

fun main() {
    val prices = listOf(1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0)
    val period = 5
    val smaResult = getSMAResult(prices, period)

    println("Period: ${smaResult.period}")
    println("Averages: ${smaResult.averages.toList()}")
}

File: src/resources/technical analysis/trend/simple moving average/simpleMovingAverage.rs
use std::iter::Iterator;

pub struct SMAResult {
    period: usize,
    averages: Box<dyn Iterator<Item = f64>>,
}

impl SMAResult {
    pub fn new(period: usize, averages: Box<dyn Iterator<Item = f64>>) -> Self {
        SMAResult { period, averages }
    }

    pub fn period(&self) -> usize {
        self.period
    }

    pub fn averages(&mut self) -> &mut dyn Iterator<Item = f64> {
        &mut self.averages
    }
}

struct SMAIterator {
    prices: Vec<f64>,
    period: usize,
    buffer: Vec<f64>,
    sum: f64,
    index: usize,
    current: usize,
}

impl SMAIterator {
    fn new(prices: Vec<f64>, period: usize) -> Self {
        let mut sum = 0.0;
        let buffer: Vec<f64> = prices.iter().take(period).map(|&x| {
            sum += x;
            x
        }).collect();

        SMAIterator {
            prices,
            period,
            buffer,
            sum,
            index: 0,
            current: 0,
        }
    }

    fn is_power_of_two(n: usize) -> bool {
        n > 0 && (n & (n - 1)) == 0
    }
}

impl Iterator for SMAIterator {
    type Item = f64;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current >= self.prices.len() - self.period + 1 {
            return None;
        }

        if self.current > 0 {
            self.sum -= self.buffer[self.index];
            self.sum += self.prices[self.current + self.period - 1];
            self.buffer[self.index] = self.prices[self.current + self.period - 1];
            self.index = (self.index + 1) % self.period;
        }

        let average = if Self::is_power_of_two(self.period) {
            self.sum * (1.0 / self.period as f64)
        } else {
            self.sum / self.period as f64
        };

        self.current += 1;
        Some(average)
    }
}

pub fn calculate_sma(prices: Vec<f64>, period: usize) -> Result<SMAResult, String> {
    if period == 0 {
        return Err("Period must be a positive integer.".to_string());
    }
    if prices.len() < period {
        return Err("Prices vector length must be greater than or equal to the period.".to_string());
    }

    let iterator = Box::new(SMAIterator::new(prices, period));
    Ok(SMAResult::new(period, iterator))
}

fn main() {
    let prices = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0];
    let period = 5;

    match calculate_sma(prices, period) {
        Ok(mut sma_result) => {
            println!("Period: {}", sma_result.period());
            println!("Averages: {:?}", sma_result.averages().collect::<Vec<f64>>());
        },
        Err(e) => println!("Error: {}", e),
    }
}